### PR TITLE
#54 Integrate auto-refresh token

### LINKED ISSUE #54: [Integrate] As a user, when my token is expired, the app will automatically refresh it.
## Why
When an access token is expired, the app should try to renew the token with the refresh_token; this would let the users continue to use the application without login in again.

## Acceptance Criteria
When the access token is nearly expired (30 seconds), the app should try to renew the token.

If it's failed, then log the user out.

### PR DESCRIPTION
**Pull request by sleepylee**
Closes #54 

## What happened 👀

Implement an automated refresh-token mechanism.
 
## Insight 📝

- Added [fresh_graphql](https://github.com/felangel/fresh/tree/master/packages/fresh_graphql) -> actually I customized the entire `Link` from them, as there are some different behavior that Exception from Unauthorized (401) request doesn't pass through the Link automatically.
- Think of `Link` like an `Interceptor` and this `CustomRefreshTokenLink` like an `Authenticator` where it will check for the response error (if exists) - then try to fetch a new token, and `forward` the old resource request with the updated tokens. 
 
## Proof Of Work 📹
To test the behavior, I added this code to change the token into an invalid token, on the next time app restarted, it refreshes correctly:

```
Future.delayed(Duration(seconds: 5), () {
      prefs.setString("PREF_KEY_TOKEN", "pR4ri-g7BueKiZ5SHi6QNc4DCHxp3uids1rotzHEdc5");
      print("Hijack token done, token is: ${prefs.getString("PREF_KEY_TOKEN")} refresh token is: ${prefs.getString("PREF_KEY_REFRESH_TOKEN")}");
    });
```

https://user-images.githubusercontent.com/13435717/127774518-0b015da3-85af-457d-8201-f1d27637fc33.mov

### CODE REVIEW

File: lib/api/graphql/graphql_client_provider.dart [hunk 1]
```diff
@@ -8,35 +15,72 @@ class GraphQLClientProvider {
...
+          final tokenResponse = refreshResult.data.authToken;
+          return OAuth2Token(
+            accessToken: tokenResponse.accessToken,
+            refreshToken: tokenResponse.refreshToken,
+            expiresIn: tokenResponse.expiresIn,
+            tokenType: tokenResponse.tokenType,
+          );
+        } else {
+          return null;
+        }
```
Comment (markgravity) [inline lib/api/graphql/graphql_client_provider.dart:72]: What do think about this suggestion? It has better readable in IMO 🤔 
```suggestion
        if (refreshResult.data == null) {
          return null;
        }
        
         final tokenResponse = refreshResult.data.authToken;
         return OAuth2Token(
            accessToken: tokenResponse.accessToken,
            refreshToken: tokenResponse.refreshToken,
            expiresIn: tokenResponse.expiresIn,
            tokenType: tokenResponse.tokenType,
          );
```

**Review by markgravity — APPROVED** [review]
LGTM with small suggestion 👍

File: lib/api/graphql/link/custom_refresh_link.dart [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+          final newRequest = request.updateContextEntry<HttpLinkHeaders>(
+            (headers) => HttpLinkHeaders(
+              headers: {
+                ...headers?.headers ?? <String, String>{},
+              }..addAll(tokenHeaders),
+            ),
+          );
+          yield* forward(newRequest);
+        } on RevokeTokenException catch (_) {
+          unawaited(revokeToken());
```
Comment (manh-t) [inline lib/api/graphql/link/custom_refresh_link.dart:123]: 👍

**Review by manh-t — APPROVED** [review]
nice documentation, thank you 💯 LGTM 🚀

File: lib/api/graphql/link/custom_refresh_link.dart [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+  @override
+  Stream<Response> request(
+    Request request, [
+    NextLink forward,
+  ]) async* {
+    final currentToken = await token;
+    final tokenHeaders = currentToken != null && _tokenHeader != null
+        ? _tokenHeader(currentToken)
+        : const <String, String>{};
```
Comment (sleepylee) [inline lib/api/graphql/link/custom_refresh_link.dart:95]: actually, I feel like I can skip the AuthLink and just pass the logic to here for the token header appending 🤔  let me recheck it a bit

Comment (sleepylee) [reply lib/api/graphql/link/custom_refresh_link.dart:95]: I was right, I will simplify the implementation and making a `Header Interceptor` + `Authenticator` on failure into this single Link, please help me review again on the next push, @manh-t @markgravity 🙇

Comment (manh-t) [reply lib/api/graphql/link/custom_refresh_link.dart]: @sleepylee it's good to go 🙏

**Pull request merged**